internal/manifests/config/configmap.go

```diff
@@ -2,172 +2,43 @@ package config
-				}{
-					Protocols: struct {
-						Grpc struct {
-							Endpoint string `yaml:"endpoint"`
-						} `yaml:"grpc"`
-					}{
-						Grpc: struct {
-							Endpoint string `yaml:"endpoint"`
-						}{
-							Endpoint: "0.0.0.0:4317",
-						},
-					},
-				},
-			},
-		},
-		Ingester: ingester.Config{
-			LifecyclerConfig: ring.LifecyclerConfig{
-				RingConfig: ring.Config{
-					KVStore: kv.Config{
-						Store: kvStore,
-					},
-					ReplicationFactor: 1,
-				},
-			},
-			MaxBlockDuration: time.Minute * 60,
-		},
-		StorageConfig: storage.Config{
-			Trace: tempodb.Config{
-				WAL: &wal.Config{
-					Filepath: "/war/tempo/wal",
-					Encoding: backend.EncSnappy,
-				},
-				Block: &common.BlockConfig{
-					IndexDownsampleBytes: 1000000,
-					IndexPageSizeBytes:   1000000,
-					BloomFP:              0.01,
-					BloomShardSizeBytes:  100000,
-					Version:              "v2",
-					Encoding:             backend.EncZstd,
-					SearchEncoding:       backend.EncSnappy,
-					SearchPageSizeBytes:  1000000,
-				},
-				Backend: "local",
-				Local: &local.Config{
-					Path: "/tmp/tempo",
-				},
-			},
-		},
-		MemberlistKV: memberlist.KVConfig{
-			JoinMembers: []string{fmt.Sprintf("tempo-%s-distributed-gossip-ring", tempo.Name)},
-		},
-	}
-
-	bytes, err := yaml.Marshal(cfg)
-	if err != nil {
-		return "", err
-	}
-	cfgStr := string(bytes)
-
-	// TODO(pavolloffay) find a nicer approach
-	// The config struct does not define omitemty which reflects an empty fields to be rendered
-	// This causes issues when Tempo is parsing the config
-	// e.g. field use_flatbuffer_search not found in type ingester.Config
-	re := regexp.MustCompile("(?m)[\r\n]+^.*tls_cipher_suites.*$")
-	cfgStr = re.ReplaceAllString(cfgStr, "")
-	re = regexp.MustCompile("(?m)[\r\n]+^.*tls_min_version.*$")
-	cfgStr = re.ReplaceAllString(cfgStr, "")
-	re = regexp.MustCompile("(?m)[\r\n]+^.*use_flatbuffer_search.*$")
-	cfgStr = re.ReplaceAllString(cfgStr, "")
-	re = regexp.MustCompile("(?m)[\r\n]+^.*metrics_ingestion_time_range_slack.*$")
-	cfgStr = re.ReplaceAllString(cfgStr, "")
-	// ingester.lifecycler.address
-	re = regexp.MustCompile("(?m)[\r\n]+^.*address.*$")
-	cfgStr = re.ReplaceAllString(cfgStr, "")
-	return cfgStr, nil
-}
-
 // Params holds configuration parameters.
 type Params struct {
-	S3 S3
-}
-
-// S3 holds S3 object storage configuration options.
-type S3 struct {
-	Endpoint string
-	Bucket   string
+	S3 S3Options
 }
 
-// BuildConfigs creates configuration objects.
-func BuildConfigs(tempo v1alpha1.Microservices, params Params) (*corev1.ConfigMap, error) {
-	s3Insecure := false
-	s3Endpoint := params.S3.Endpoint
-	if strings.HasPrefix(s3Endpoint, "http://") {
-		s3Insecure = true
-		s3Endpoint = strings.TrimPrefix(s3Endpoint, "http://")
-	} else if !strings.HasPrefix(s3Endpoint, "https://") {
-		s3Insecure = true
+// BuildConfigMap creates configuration objects.
+func BuildConfigMap(tempo v1alpha1.Microservices, params Params) (*corev1.ConfigMap, error) {
```

Comment:
It seems that we are leaking some APIs or abstractions here. 

```go
// Params holds configuration parameters.
type Params struct {
	S3 S3Options
}

// S3 holds S3 object storage configuration options.
type S3Options struct {
	Endpoint string
	Bucket   string
	Insecure bool
}

// Options holds the configuration template options.
type Options struct {
	GlobalRetention        string
	QueryFrontendDiscovery string
	MemberList             []string
	S3                     S3Options
}
```

I don't like that `Params` encapsulate `S3Options` that are used in `Options` for the rendering. Could be `Options` defined as private (with all it's dependencies)?

Reply:
Done